File: src/Board.java

public class Board {
	
	private int board[][];
	public final int width = 7;
	public final int height = 6;
	public int score = 1000;
	
	public void printBoard(){
		System.out.println();
		for(int i =0 ; i < this.height; i++){
			int row[] = this.board[i];
			for(int j = 0; j < this.width; j++){
				System.out.print(row[j] + ", ");
			}
			System.out.println();
		}
		System.out.println();
	}
	
	public int[][] getBoard(){
		int [][] myInt = new int[this.height][];
		for(int i = 0; i < this.height; i++)
		{
		  int[] aMatrix = this.board[i];
		  int   aLength = aMatrix.length;
		  myInt[i] = new int[aLength];
		  System.arraycopy(aMatrix, 0, myInt[i], 0, aLength);
		}
		
		return myInt;
	}
	
	public void initBoard(){
		for(int i =0 ; i < this.height; i++){
			int row[] = this.board[i];
			for(int j = 0; j < this.width; j++){
				row[j] = 0;
			}
		}
	}
	
	public Board(){
		board = new int[this.height][this.width];
		this.initBoard();
		this.printBoard();
	}
	
	public Board(Board b){
		board = new int[6][7];
		this.board = b.getBoard().clone();
	}
	
	public boolean addChip(int player, int column){
		if(column < 0 || column >= this.width){
			return false;
		}
		for(int i = (this.height - 1); i >= 0; i--){
			if(board[i][column] == 0){
				board[i][column] = player;
				return true;
			}
		}
		return false;
	}
	
	public boolean isFull(){
		for (int i=this.height-1; i>=0;i--){
			for(int j=this.width-1;j>=0;j--){
				if(this.board[i][j]==0){
					return false;
				}
			}
		}
		return true;
	}
	
	public boolean isFinished(int depth, int score) {
	    if (depth == 0 || score == this.score || score == -this.score || this.isFull()) {
	        return true;
	    }
	    return false;
	}
	
	public int getCurrentScore(){
		int[][] Board = this.getBoard();
		int length = this.height;
		int wide = this.width;
		int cpu = 100;
		int player = -100;
		int auxP=0;
		int auxC=0;
		for (int i=length-1; i>=0;i--){
			for(int j=wide-1;j>=0;j--){
			if(Board[i][j]==1){
				auxP++;
				auxC=0;
			}else if (Board[i][j]==2){
				auxC++;
				auxP=0;
			}else{
				auxC=0;
				auxP=0;
			}
			if(auxC==4){
				System.out.println("gano cpu en ");
				return cpu;
			}
			if(auxP==4){
				return player;
			}
			}
			auxC=0;
			auxP=0;
		}
		auxC=0;
		auxP=0;
		for(int j=wide-1;j>=0;j--){
			for (int i=length-1; i>=0;i--){
				if(Board[i][j]==1){
					auxP++;
					auxC=0;
				}else if (Board[i][j]==2){
					auxC++;
					auxP=0;
				}else{
					auxC=0;
					auxP=0;
				}
				if(auxC==4){
					return cpu;
				}
				if(auxP==4){
					return player;
				}
				}
			auxC=0;
			auxP=0;
				
			}
		
		auxC=0;
		auxP=0;
		for(int i=0;i<wide-1;i++){
	        for(int j=0;j<length-1;j++){
	        	
	        	int j3=j+3;
	        	int i3= i+3;
	        	
	        	if(i3<wide-1 && j3 <length-1){
	        	
	        	
	            if(Board[i][j] == 1 && Board[i][j]==Board[i+1][j+1] && Board[i][j]==Board[i+2][j+2] && Board[i][j]==Board[i+3][j+3])
	            	return player;
	            else if(Board[i][j] == 2 && Board[i][j]==Board[i+1][j+1] && Board[i][j]==Board[i+2][j+2] && Board[i][j]==Board[i+3][j+3]){
	            	System.out.println("entro");
	            	return cpu;
	            }
	        	}
	        }
	        
		}
	    //checks left diagonal win
		for(int i=0;i<wide-1;i++){
	        for(int j=0;j<length-1;j++){
	        	int j3=j-3;
	        	int i3= i+3;
	        	int ja3=j+3;
	        	int ia3=i-3;
	        	if(i3<wide-1 && j3 >0 ){
	        	
	            if(Board[i][j] == 1 && Board[i][j]==Board[i+1][j-1] && Board[i][j]==Board[i+2][j-2] && Board[i][j]==Board[i+3][j-3])
	            	return player;
	            else if(Board[i][j] == 2 && Board[i][j]==Board[i+1][j-1] && Board[i][j]==Board[i+2][j-2] && Board[i][j]==Board[i+3][j-3])
	            	return cpu;
	            
	          
	            
	        	}
	        	
	        	if(ia3>0 &&ja3<length-1){
	        		  if(Board[i][j] == 1 && Board[i][j]==Board[i-1][j+1] && Board[i-1][j+1]==Board[i-2][j+2] && Board[i][j]==Board[i-3][j+3])
	  	            	return player;
	  	            else if(Board[i][j] == 2 && Board[i][j]==Board[i-1][j+1] && Board[i-1][j+1]==Board[i-2][j+2] && Board[i][j]==Board[i-3][j+3])
	  	            	return cpu;
	        		
	        	}
	        	
	        	
	        	
	     
	        } 	
		}
		return 0;
	}
	
	public int scorePosition (int row, int column, int delta_y, int delta_x) {
	    int human_points = 0;
	    int computer_points = 0;
	    int[][] board = this.getBoard();

	    for (int i = 0; i < 4; i++) {
	        if (board[row][column] == 1) {
	            human_points++; 
	        } else if (board[row][column] == 2) {
	            computer_points++; 
	        }

	        row += delta_y;
	        column += delta_x;
	    }

	    if (human_points == 4) {
	        return -this.score;
	    } else if (computer_points == 4) {
	        return this.score;
	    } else {
	        return computer_points;
	    }
	}
	
	
	public int getScore() {
	    int points = 0;

	    int vertical_points = 0;
	    int horizontal_points = 0;
	    int diagonal_points1 = 0;
	    int diagonal_points2 = 0;


	    for (int row = 0; row < this.height - 3; row++) {
	        for (int column = 0; column < this.width; column++) {
	            int score = this.scorePosition(row, column, 1, 0);
	            if (score == this.score){
	            	return this.score;
	            }
	            if (score == -this.score){
	            	return -this.score;
	            }
	            vertical_points += score;
	        }            
	    }

	    for (int row = 0; row < this.height; row++) {
	        for (int column = 0; column < this.width - 3; column++) {
	            int score = this.scorePosition(row, column, 0, 1);   
	            if (score == this.score){
	            	return this.score;
	            }
	            if (score == -this.score){
	            	return -this.score;
	            }
	            horizontal_points += score;
	        } 
	    }

	    for (int row = 0; row < this.height - 3; row++) {
	        for (int column = 0; column < this.width - 3; column++) {
	            int score = this.scorePosition(row, column, 1, 1);
	            if (score == this.score){
	            	return this.score;
	            }
	            if (score == -this.score){
	            	return -this.score;
	            }
	        }            
	    }

	    for (int row = 3; row < this.height; row++) {
	        for (int column = 0; column < this.width - 4; column++) {
	            int score = this.scorePosition(row, column, -1, +1);
	            if (score == this.score){
	            	return this.score;
	            }
	            if (score == -this.score){
	            	return -this.score;
	            }
	            diagonal_points2 += score;
	        }

	    }

	    points = horizontal_points + vertical_points + diagonal_points1 + diagonal_points2;
	    return points;
	}
}

File: conecta_4_python/class/Board.py
class Board:
    board = None;
    columns = 7;
    rows = 6;
    MAX_SCORE = 10000;


    def printBoard(self):
        print("\n");
        for row in range(0, self.rows):
            for column in range(0, self.columns):
                print(str(self.board[row][column]) + ", ")
		print("\n");


    def initBoard(self):
        self.board = [[]];
        for row in range(0,self.rows):
            for column in range(0,self.columns):
                self.board[row][column] = 0;

    def __init__(self):
        self.initBoard();

    def isFinished(self):
        if self.isFull or abs(self.getScore) > 1000:
            return True;
        return False;

    def addChip(self, player, column):
        if column >= 0 and column <= self.columns:
            return false;
        for row in range(self.rows - 1 , 0):
            if self.board[row][column] == 0:
                board[row][column] = player;
                return true;
        return false;

    def isFull(self):
        for row in range(0,self.rows):
            for column in range(0,self.columns):
                if self.board[row][row] == 0:
                    return false
        return true

    def getVerticalPoints(self, row, column):
        oponentPoints = 0;
        myPoints = 0;
        for  i in range(0,4):
            if self.board[row][column] == 1:
                oponentPoints++;
            else if self board[row][column] == 2:
                myPoints++
            row += 1

        if oponentPoints == 4:
            return -self.MAX_SCORE;
        if myPoints == 4:
            return self.MAX_SCORE;

        return computer_points;

    def getHorizontalPoints(self, row, column):
        oponentPoints = 0;
        myPoints = 0;

        for  i in range(0,4):
            if self.board[row][column] == 1:
                oponentPoints++;
            else if self board[row][column] == 2:
                myPoints++
            column += 1

        if oponentPoints == 4:
            return -self.MAX_SCORE;
        if myPoints == 4:
            return self.MAX_SCORE;

        return computer_points;

    def getDiagonalLeftPoints(self, row, column):
        oponentPoints = 0;
        myPoints = 0;

        for  i in range(0,4):
            if self.board[row][column] == 1:
                oponentPoints++;
            else if self board[row][column] == 2:
                myPoints++
            column += 1
            row +=1

        if oponentPoints == 4:
            return -self.MAX_SCORE;
        if myPoints == 4:
            return self.MAX_SCORE;

        return computer_points;

    def getDiagonalRightPoints(self, row, column):
        oponentPoints = 0;
        myPoints = 0;

        for  i in range(0,4):
            if self.board[row][column] == 1:
                oponentPoints++;
            else if self board[row][column] == 2:
                myPoints++
            column -= 1
            row +=1

        if oponentPoints == 4:
            return -10000;
        if myPoints == 4:
            return 10000;

        return computer_points;

    def getScore(self):
        points = 0;

        verticalPoints = 0;
        horizontalPoints = 0;
        diagonalLeftPoint = 0;
        diagonalRightPoints = 0;

        for row in range(0, self.height - 3):
            for column in range(0, self.width):
                score = this.getVerticalPoints(row, column);
                if abs(score) == self.MAX_SCORE
                    return score
                verticalPoints += score;


        for row in range(0, self.height):
            for column in range(0, self.width - 3):
                score = this.getHorizontalPoints(row, column);
                if abs(score) == self.MAX_SCORE
                    return score
                horizontalPoints += score;

        for row in range(0, self.height - 3):
            for column in range(0, self.width - 3):
                score = this.getDiagonalLeftPoints(row, column);
                if abs(score) == self.MAX_SCORE
                    return score
                diagonalLeftPoint += score;

        for row in range(3, self.height):
            for column in range(0, self.width - 4):
                score = this.getDiagonalLeftPoints(row, column);
                if abs(score) == self.MAX_SCORE
                    return score
                diagonalRightPoints += score;

        return verticalPoints + horizontalPoints + diagonalLeftPoint + diagonalRightPoints;
